jetty-server/src/main/java/org/eclipse/jetty/server/session/AbstractSessionCache.java

```diff
@@ -501,6 +519,48 @@ public void add(String id, Session session) throws Exception
         }
     }
 
+    /**
+     * A response that has accessed this session is about to
+     * be returned to the client. Pass the session to the store
+     * to persist, so that any changes will be visible to
+     * subsequent requests on the same node (if using NullSessionCache),
+     * or on other nodes.
+     */
+    @Override
+    public void commit(Session session) throws Exception
+    {
+        if (session == null)
+            return;
+
+        try (Lock lock = session.lock())
+        {
+            //only write the session out at this point if the attributes changed. If only
+            //the lastAccess/expiry time changed defer the write until the last request exits
+            if (session.getSessionData().isDirty() && _flushOnResponseCommit)
+            {
+                //save the session
+                if (!_sessionDataStore.isPassivating())
+                {
+                    _sessionDataStore.store(session.getId(), session.getSessionData());
+                }
+                else
+                {
+                    //A passivate listener might remove a non-serializable attribute that
+                    //the activate listener might put back in again, which would spuriously
+                    //set the dirty bit to true, causing another round of passivate/activate
+                    //when the request exits. The store clears the dirty bit if it does a
+                    //save, so ensure dirty flag is set to the value determined by the store,
+                    //not a passivation listener.
+                    session.willPassivate();
+                    _sessionDataStore.store(session.getId(), session.getSessionData());
+                    boolean dirty = session.getSessionData().isDirty();
+                    session.didActivate();
+                    session.getSessionData().setDirty(dirty);
```

Comment:
Can these dirty logic in the 3 lines above be moved into `session.didActivate` so they do not have to be duplicated elsewhere?   Is it ever appropriate to call `didActivate` without the dirty bit logic?